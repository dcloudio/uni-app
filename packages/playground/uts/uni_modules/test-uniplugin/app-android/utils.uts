export function test(){
    console.log('test')
}