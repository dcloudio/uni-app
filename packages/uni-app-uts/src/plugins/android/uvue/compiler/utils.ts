import path from 'path'
import type { Node } from '@babel/types'
import { type ExpressionNode, createSimpleExpression } from '@vue/compiler-core'
import { walk } from 'estree-walker'
import { parseExpression } from '@babel/parser'
import MagicString from 'magic-string'
import {
  camelize,
  genUTSComponentPublicInstanceIdent,
  normalizePath,
} from '@dcloudio/uni-cli-shared'
import type { TemplateCompilerOptions } from './options'
import { stringifyExpression } from './transforms/transformExpression'
import type { TransformContext } from './transform'
import type { CompilerError } from './errors'

export const __DEV__ = true
export const __BROWSER__ = false
export const __COMPAT__ = false

export function isCompatEnabled(...args: any[]) {
  return false
}

export function genRenderFunctionDecl({
  className,
  genDefaultAs,
  inline = false,
}: TemplateCompilerOptions & { genDefaultAs?: string }): string {
  if (inline) {
    return `(): any | null =>`
  }
  // 调整返回值类型为 any | null, 支持 <template>some text</template>
  const thisCode = genDefaultAs
    ? `this: InstanceType<typeof ${genDefaultAs}>`
    : ''
  return `function ${className || ''}Render(${thisCode}): any | null`
}

export function rewriteObjectExpression(
  exp: ExpressionNode,
  context: TransformContext
) {
  const source = stringifyExpression(exp)
  if (source.includes('{')) {
    const s = new MagicString(source)
    const ast = parseExpression(source, {
      plugins: context.expressionPlugins,
    })
    walk(ast, {
      enter(node: Node) {
        if (node.type === 'ObjectExpression') {
          s.prependLeft(
            node.start!,
            node.properties.length > 0 ? '_uM(' : '_uM<string, any | null>('
          )
          s.prependRight(node.end!, ')')
        }
      },
    })
    return createSimpleExpression(s.toString(), false, exp.loc)
  }
}

export function onCompilerError(error: CompilerError) {}

export function parseSource(fileName: string, rootDir: string) {
  if (fileName.includes('@dcloudio')) {
    return fileName
  }
  rootDir = normalizePath(rootDir)
  if (path.isAbsolute(fileName) && fileName.startsWith(rootDir)) {
    return '@/' + normalizePath(path.relative(rootDir, fileName))
  }
  return fileName
}

export function addEasyComponentAutoImports(
  easyComponentAutoImports: Record<string, [string, string]>,
  rootDir: string,
  tagName: string,
  fileName: string
) {
  if (easyComponentAutoImports[fileName]) {
    return
  }
  // 内置easycom，如 unicloud-db
  if (fileName.includes('@dcloudio')) {
    return
  }
  rootDir = normalizePath(rootDir)
  if (path.isAbsolute(fileName) && fileName.startsWith(rootDir)) {
    fileName = '@/' + normalizePath(path.relative(rootDir, fileName))
  }
  // 加密插件easycom类型导入
  if (fileName.includes('?uts-proxy')) {
    const moduleId = path.basename(fileName.split('?uts-proxy')[0])
    fileName = `uts.sdk.modules.${camelize(moduleId)}`
  }
  const ident = genUTSComponentPublicInstanceIdent(tagName)
  easyComponentAutoImports[fileName] = [ident, ident]
}
