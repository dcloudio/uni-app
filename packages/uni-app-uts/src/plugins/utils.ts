import path from 'node:path'
import {
  isAppHarmonyUVueNativeTag,
  isAppIOSUVueNativeTag,
  isAppUVueBuiltInEasyComponent,
} from '@dcloudio/uni-shared'
import {
  MANIFEST_JSON_UTS,
  PAGES_JSON_UTS,
  type UniViteCopyPluginOptions,
  type UniVitePlugin,
  initI18nOptions,
  injectAssetPlugin,
  matchUTSComponent,
  normalizeNodeModules,
  normalizePath,
  parseUTSComponent,
  removePlugins,
  transformTapToClick,
  transformUTSComponent,
} from '@dcloudio/uni-cli-shared'
import { compileI18nJsonStr } from '@dcloudio/uni-i18n'
import type { ResolvedConfig } from 'vite'
import { ElementTypes, NodeTypes } from '@vue/compiler-core'

export function createUniOptions(
  platform: 'app-android' | 'app-ios' | 'app-harmony'
): UniVitePlugin['uni'] {
  return {
    copyOptions() {
      const inputDir = process.env.UNI_INPUT_DIR
      const outputDir = process.env.UNI_OUTPUT_DIR
      const targets: UniViteCopyPluginOptions['targets'] = []
      // 自动化测试时，不启用隐私政策
      if (!process.env.UNI_AUTOMATOR_WS_ENDPOINT) {
        if (process.env.UNI_UTS_PLATFORM === 'app-android') {
          targets.push({
            src: 'androidPrivacy.json',
            dest: outputDir,
            transform(source) {
              const options = initI18nOptions(
                process.env.UNI_PLATFORM,
                inputDir,
                false,
                true
              )
              if (!options) {
                return
              }
              return compileI18nJsonStr(source.toString(), options)
            },
          })
        }
      }
      return {
        assets: ['hybrid/html/**/*', 'uni_modules/*/hybrid/html/**/*'],
        targets,
      }
    },
    compilerOptions:
      platform === 'app-ios' || platform === 'app-harmony'
        ? {
            isNativeTag(tag) {
              return (
                matchUTSComponent(tag) ||
                (platform === 'app-ios'
                  ? isAppIOSUVueNativeTag
                  : isAppHarmonyUVueNativeTag)(tag)
              )
            },
            nodeTransforms: [
              transformTapToClick,
              transformUTSComponent,
              // TODO 合并复用安卓插件逻辑
              function (node, context) {
                if (node.type === 2) {
                  const parent = context.parent
                  if (parent && parent.type === 1 && parent.tag === 'text') {
                    // 解析文本节点转义，暂时仅处理换行
                    node.content = node.content.replace(
                      /[\\]+n/g,
                      function (match) {
                        return JSON.parse(`"${match}"`)
                      }
                    )
                  }
                }
              },
              (node) => {
                // 收集可能的 extApiComponents
                if (
                  node.type === NodeTypes.ELEMENT &&
                  (node.tagType === ElementTypes.ELEMENT ||
                    (node.tagType === ElementTypes.COMPONENT &&
                      isAppUVueBuiltInEasyComponent(node.tag)))
                ) {
                  if (!parseUTSComponent(node.tag, 'swift')) {
                    addExtApiComponents([node.tag])
                  }
                }
              },
            ],
          }
        : {},
  }
}

export function isManifest(id: string) {
  return id.endsWith(MANIFEST_JSON_UTS)
}

export function isPages(id: string) {
  return id.endsWith(PAGES_JSON_UTS)
}

// TODO vite 升级需要考虑调整以下列表
const REMOVED_PLUGINS = [
  'vite:build-metadata',
  'vite:modulepreload-polyfill',
  // 'vite:css', // iOS replace
  'vite:esbuild',
  'vite:wasm-helper',
  'vite:worker',
  // 'vite:json',
  // 'vite:asset', // replace
  'vite:wasm-fallback',
  // 'vite:define',
  // 'vite:css-post', // iOS replace
  'vite:build-html',
  'vite:html-inline-proxy',
  'vite:worker-import-meta-url',
  'vite:asset-import-meta-url',
  'vite:force-systemjs-wrap-complete',
  'vite:watch-package-data',
  'commonjs',
  'vite:data-uri',
  'vite:dynamic-import-vars',
  'vite:import-glob',
  'vite:build-import-analysis',
  'vite:terser',
  'vite:reporter',
]

if (process.env.UNI_UTS_PLATFORM === 'app-android') {
  REMOVED_PLUGINS.push('vite:esbuild-transpile')
}

export function configResolved(config: ResolvedConfig, isAndroidX = false) {
  removePlugins(REMOVED_PLUGINS.slice(0), config)
  // console.log(plugins.map((p) => p.name))
  // 强制不inline
  config.build.assetsInlineLimit = 0
  injectAssetPlugin(config, { isAndroidX })
}

export function relativeInputDir(filename: string) {
  const inputDir = process.env.UNI_INPUT_DIR
  filename = normalizeNodeModules(filename)
  if (filename.startsWith(inputDir)) {
    return normalizePath(path.relative(inputDir, filename))
  }
  return filename
}

export function normalizeManifestJson(userManifestJson: Record<string, any>) {
  const app = userManifestJson.app || {}
  const x = userManifestJson['uni-app-x'] || {}
  x.compilerVersion = process.env.UNI_COMPILER_VERSION || ''
  const pageOrientation = getGlobalPageOrientation()

  if (pageOrientation) {
    if (!app.distribute) {
      app.distribute = {}
    }
    app.distribute['_uni-app-x_'] = {
      pageOrientation,
    }
    app.distribute.globalStyle = {
      pageOrientation,
    }
  }

  return {
    id: userManifestJson.appid || '',
    name: userManifestJson.name || '',
    description: userManifestJson.description || '',
    version: {
      name: userManifestJson.versionName || '',
      code: userManifestJson.versionCode || '',
    },
    'uni-app-x': x,
    app,
  }
}

export function updateManifestModules(
  manifest: Record<string, any>,
  modules: string[]
) {
  // 执行了摇树逻辑，就需要设置 modules 节点
  const app = manifest.app
  if (!app.distribute) {
    app.distribute = {}
  }
  if (!app.distribute.modules) {
    app.distribute.modules = {}
  }
  if (modules) {
    modules.forEach((name) => {
      const value = app.distribute.modules[name]
      app.distribute.modules[name] =
        typeof value === 'object' && value !== null ? value : {}
    })
  }
  return manifest
}

let pageOrientation = ''

export function setGlobalPageOrientation(value: string) {
  pageOrientation = value
}

export function getGlobalPageOrientation() {
  return pageOrientation
}

const extApiComponents: Set<string> = new Set()
export function addExtApiComponents(components: string[]) {
  components.forEach((component) => {
    extApiComponents.add(component)
  })
}

export function getExtApiComponents() {
  return extApiComponents
}
