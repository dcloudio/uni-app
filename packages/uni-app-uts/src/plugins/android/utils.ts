import fs from 'fs'
import path from 'path'
import { type ImportSpecifier, init, parse } from 'es-module-lexer'
import {
  createResolveErrorMsg,
  createRollupError,
  getUTSEasyComAutoImports,
  normalizeNodeModules,
  offsetToStartAndEnd,
  parseUniExtApiNamespacesJsOnce,
} from '@dcloudio/uni-cli-shared'
import { isArray, isPlainObject, isString } from '@vue/shared'

import { type Import, createUnimport } from 'unimport'

import type { /*SourceMapInput, */ PluginContext } from 'rollup'
import type { Position, SourceLocation } from '@vue/compiler-core'

import { createCompilerError } from './uvue/compiler/errors'

export const UVUE_CLASS_NAME_PREFIX = 'Gen'

export const DEFAULT_APPID = '__UNI__uniappx'

export const ENTRY_FILENAME = () =>
  process.env.UNI_APP_X_TSC === 'true' ? 'main.uts.ts' : 'main.uts'

export function wrapResolve(
  resolve: PluginContext['resolve']
): PluginContext['resolve'] {
  return async (source, importer, options) => {
    try {
      return await resolve(source, importer, options)
    } catch (e) {
      // import "@/pages/logo.png" 可能会报 Cannot find module 错误
    }
    return null
  }
}

export function createTryResolve(
  importer: string,
  resolve: PluginContext['resolve'],
  offsetStart?: Position,
  origCode: string = ''
) {
  return async (
    source: string,
    code: string,
    { ss, se }: ImportSpecifier
  ): Promise<boolean | void> => {
    const resolved = await wrapResolve(resolve)(source, importer)
    if (!resolved) {
      const { start, end } = offsetToStartAndEnd(code, ss, se)
      if (offsetStart) {
        if (start.line === 1) {
          start.column = start.column + offsetStart.column
          if (end.line === 1) {
            end.column = end.column + offsetStart.column
          }
        }
        const offsetLine = offsetStart.line - 1
        start.line = start.line + offsetLine
        end.line = end.line + offsetLine
      }
      throw createResolveError(
        origCode || code,
        createResolveErrorMsg(source, importer),
        start,
        end
      )
    }
  }
}

export async function parseImports(
  code: string,
  tryResolve?: ReturnType<typeof createTryResolve>
) {
  await init
  let res: ReturnType<typeof parse> = [[], [], false, false]
  try {
    res = parse(code)
  } catch (err: any) {
    const message = err.message
    if (message) {
      const matches = message.match(/@:(\d+):(\d+)/)
      if (matches) {
        throw createRollupError(
          '',
          '',
          createCompilerError(
            0,
            {
              start: {
                offset: 0,
                line: parseInt(matches[1]),
                column: parseInt(matches[2]),
              },
            } as SourceLocation,
            { 0: `Parse error` },
            ''
          ),
          code
        )
      }
    }
    throw err
  }
  const imports = res[0]
  const importsCode: string[] = []
  for (const specifier of imports) {
    const source = code.slice(specifier.s, specifier.e)
    if (tryResolve) {
      const res = await tryResolve(source, code, specifier)
      if (res === false) {
        // 忽略该import
        continue
      }
    }
    importsCode.push(`import "${source}"`)
  }

  return importsCode.concat(parseUniExtApiImports(code)).join('\n')
}

export function createResolveError(
  code: string,
  msg: string,
  start: Position,
  end: Position
) {
  return createRollupError(
    '',
    '',
    createCompilerError(
      0,
      {
        start,
        end,
      } as SourceLocation,
      { 0: msg },
      ''
    ),
    code
  )
}

// @ts-expect-error 暂时不用
function genImportsCode(code: string, imports: readonly ImportSpecifier[]) {
  const chars = code.split('')
  const keepChars: number[] = []
  imports.forEach(({ ss, se }) => {
    for (let i = ss; i <= se; i++) {
      keepChars.push(i)
    }
  })
  for (let i = 0; i < chars.length; i++) {
    if (!keepChars.includes(i)) {
      const char = chars[i]
      if (char !== '\r' && char !== '\n') {
        chars[i] = ' '
      }
    }
  }
  return chars.join('')
}

function parseUniExtApiImports(code: string): string[] {
  if (!process.env.UNI_UTS_PLATFORM) {
    return []
  }
  const extApis = parseUniExtApiNamespacesJsOnce(
    process.env.UNI_UTS_PLATFORM,
    process.env.UNI_UTS_TARGET_LANGUAGE
  )
  const pattern = /uni\.(\w+)/g
  const apis = new Set<string>()
  let match
  while ((match = pattern.exec(code)) !== null) {
    apis.add(match[1])
  }
  const imports: string[] = []
  apis.forEach((api) => {
    const extApi = extApis[api]
    if (extApi) {
      imports.push(`import "${extApi[0]}"`)
    }
  })
  return imports
}

export function kotlinOutDir() {
  return path.join(process.env.UNI_OUTPUT_DIR, '../.kotlin')
}

export function isVue(filename: string) {
  return filename.endsWith('.vue') || filename.endsWith('.uvue')
}

export function stringifyMap(obj: unknown, ts = false) {
  return serialize(obj, ts)
}

function serialize(obj: unknown, ts = false): string {
  if (isString(obj)) {
    return `"${obj}"`
  } else if (isPlainObject(obj)) {
    const entries = Object.entries(obj).map(
      ([key, value]) => `[${serialize(key, ts)},${serialize(value, ts)}]`
    )
    if (entries.length) {
      return `utsMapOf([${entries.join(',')}])`
    }
    if (ts) {
      return `utsMapOf<string, any | null>()`
    }
    return `utsMapOf()`
  } else if (isArray(obj)) {
    return `[${obj.map((item) => serialize(item, ts)).join(',')}]`
  } else {
    return String(obj)
  }
}

export function parseUTSRelativeFilename(filename: string, root?: string) {
  if (!path.isAbsolute(filename)) {
    return filename
  }
  return normalizeNodeModules(
    path.relative(root ?? process.env.UNI_INPUT_DIR, filename)
  )
}

export function parseUTSImportFilename(filename: string) {
  if (!path.isAbsolute(filename)) {
    return filename
  }
  return (
    '@/' +
    normalizeNodeModules(path.relative(process.env.UNI_INPUT_DIR, filename))
  )
}

type UniCloudSpace = {
  provider: string
  spaceName: string
  spaceId: string
  clientSecret?: string
  endpoint?: string
  workspaceFolder?: string
}

let uniCloudSpaceList: Array<UniCloudSpace>
export function getUniCloudSpaceList(): Array<UniCloudSpace> {
  if (uniCloudSpaceList) {
    return uniCloudSpaceList
  }
  if (!process.env.UNI_CLOUD_SPACES) {
    uniCloudSpaceList = []
    return uniCloudSpaceList
  }
  try {
    const spaces = JSON.parse(process.env.UNI_CLOUD_SPACES)
    if (!Array.isArray(spaces)) {
      uniCloudSpaceList = []
      return uniCloudSpaceList
    }
    uniCloudSpaceList = spaces.map((space) => {
      if (space.provider === 'tcb') {
        space.provider = 'tencent'
      }
      if (!space.provider && space.clientSecret) {
        space.provider = 'aliyun'
      }
      switch (space.provider) {
        case 'aliyun':
        case 'dcloud':
          return {
            provider: space.provider || 'aliyun',
            spaceName: space.name,
            spaceId: space.id,
            clientSecret: space.clientSecret,
            endpoint: space.apiEndpoint,
            workspaceFolder: space.workspaceFolder,
          }
        case 'alipay': {
          return {
            provider: space.provider,
            spaceName: space.name,
            spaceId: space.id,
            spaceAppId: space.spaceAppId,
            accessKey: space.accessKey,
            secretKey: space.secretKey,
            workspaceFolder: space.workspaceFolder,
          }
        }
        case 'tencent':
        default: {
          return {
            provider: space.provider,
            spaceName: space.name,
            spaceId: space.id,
            workspaceFolder: space.workspaceFolder,
          }
        }
      }
    })
  } catch (e) {
    console.error(e)
  }
  uniCloudSpaceList = uniCloudSpaceList || []
  if (uniCloudSpaceList.length > 1) {
    console.warn('Multi uniCloud space is not supported yet.')
  }
  return uniCloudSpaceList
}

// TODO copy from uni-uts-v1 compiler, should be refactor
type UniCloudObjectInfo = {
  name: string
  methodList: string[]
}

export function getUniCloudObjectInfo(
  uniCloudSpaceList: Array<UniCloudSpace>
): Array<UniCloudObjectInfo> {
  let uniCloudWorkspaceFolder = process.env.UNI_INPUT_DIR.endsWith('src')
    ? path.resolve(process.env.UNI_INPUT_DIR, '..')
    : process.env.UNI_INPUT_DIR
  let serviceProvider = 'aliyun'
  if (uniCloudSpaceList && uniCloudSpaceList.length > 0) {
    const space = uniCloudSpaceList[0]
    if (space.workspaceFolder) {
      uniCloudWorkspaceFolder = space.workspaceFolder
    }
    serviceProvider = space.provider === 'tencent' ? 'tcb' : space.provider
  } else {
    serviceProvider =
      ['aliyun', 'tcb', 'alipay', 'dcloud'].find((item) =>
        fs.existsSync(path.resolve(uniCloudWorkspaceFolder, 'uniCloud-' + item))
      ) || 'aliyun'
  }
  try {
    const { getWorkspaceObjectInfo } = require('../../../lib/unicloud-utils')
    return getWorkspaceObjectInfo(uniCloudWorkspaceFolder, serviceProvider)
  } catch (e) {
    console.error((e as Error).message)
    process.exit(1)
  }
}

export async function transformAutoImport(
  code: string,
  id: string,
  ignore: string[] = []
) {
  const { matchedImports } = await detectAutoImports(code, id, ignore)
  if (matchedImports.length) {
    return {
      code: code + '\n' + genAutoImportsCode(matchedImports),
    }
  }
  return { code }
}

export function genAutoImportsCode(imports: Import[]) {
  const codes: string[] = []
  imports.forEach(({ name, as, from }) => {
    if (as && name !== as) {
      codes.push(
        `import { ${name} as ${as} } from "${parseUTSImportFilename(from)}"`
      )
    } else {
      codes.push(`import { ${name}  } from "${parseUTSImportFilename(from)}"`)
    }
  })
  return codes.join('\n')
}

let detectImports: DetectImports

export function transformUniCloudMixinDataCom(code: string) {
  // 将 uniCloud.mixinDatacom 替换为 uniCloudMixinDatacom
  // 然后 autoImport 会自动导入 uniCloudMixinDatacom
  if (code.includes('uniCloud.mixinDatacom')) {
    return code.replace(/uniCloud\.mixinDatacom/g, 'uniCloudMixinDatacom')
  }
  return code
}

export function detectAutoImports(
  code: string,
  id: string,
  ignore: string[] = []
) {
  // 目前硬编码
  if (id.includes('index.module.uts')) {
    return { matchedImports: [] }
  }
  if (!detectImports) {
    detectImports = initAutoImport().detectImports
  } else {
    const autoImports = getUTSEasyComAutoImports()
    const sources = Object.keys(autoImports)
    if (detectImports.key !== sources.sort().join(',')) {
      detectImports = initAutoImport().detectImports
    }
  }

  return detectImports(code, id, ignore)
}

type DetectImports = {
  key: string
  (code: string, id: string, ignore: string[]): Promise<{
    matchedImports: Import[]
  }>
}

function initAutoImport(): {
  detectImports: DetectImports
} {
  const autoImports = getUTSEasyComAutoImports()
  const sources = Object.keys(autoImports)
  if (!sources.length) {
    const detectImports = async (
      _code: string,
      _id: string,
      _ignore: string[] = []
    ) => {
      return { matchedImports: [] }
    }
    detectImports.key = 'default'
    return {
      detectImports,
    }
  }

  const imports: Import[] = []
  sources.forEach((source) => {
    autoImports[source].forEach(([name, as]) => {
      imports.push({
        name,
        as,
        from: source,
      })
    })
  })
  const { detectImports: uniDetectImports } = createUnimport({
    imports,
  })
  const detectImports = async function (
    code: string,
    id: string,
    ignore: string[] = []
  ) {
    // const start = Date.now()
    const result = await uniDetectImports(code)
    // console.log('detectImports[' + id + ']耗时:' + (Date.now() - start))
    return {
      matchedImports: result.matchedImports.filter((item) => {
        if (item.as && item.name !== item.as) {
          return !ignore.includes(item.as)
        }
        return !ignore.includes(item.name)
      }),
    }
  }
  detectImports.key = sources.sort().join(',')
  return {
    detectImports,
  }
}
