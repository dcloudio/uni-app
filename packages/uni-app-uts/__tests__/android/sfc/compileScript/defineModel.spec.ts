import { BindingTypes } from '@vue/compiler-core'
import { assertCode, compileSFCScript as compile } from '../utils'

describe('defineModel()', () => {
  test('basic usage', () => {
    const { content, bindings } = compile(
      `
      <script setup>
      const modelValue = defineModel({ required: true })
      const c = defineModel('count')
      const toString = defineModel('toString', { type: Function })
      </script>
      `,
      { defineModel: true }
    )
    assertCode(content)
    expect(content).toMatch('props: {')
    expect(content).toMatch('"modelValue": { required: true },')
    expect(content).toMatch('"count": {},')
    expect(content).toMatch('"toString": { type: Function },')
    expect(content).toMatch(
      'emits: ["update:modelValue", "update:count", "update:toString"],'
    )
    expect(content).toMatch(
      `const modelValue = useModel<any>(__ins.props, "modelValue")`
    )
    expect(content).toMatch(`const c = useModel<any>(__ins.props, "count")`)
    // expect(content).toMatch(`return { modelValue, c, toString }`)
    expect(content).not.toMatch('defineModel')

    expect(bindings).toStrictEqual({
      modelValue: BindingTypes.SETUP_REF,
      count: BindingTypes.PROPS,
      c: BindingTypes.SETUP_REF,
      toString: BindingTypes.SETUP_REF,
    })
  })

  test('w/ defineProps and defineEmits', () => {
    const { content, bindings } = compile(
      `
      <script setup>
      defineProps({ foo: String })
      defineEmits(['change'])
      const count = defineModel({ default: 0 })
      </script>
    `,
      { defineModel: true }
    )
    assertCode(content)
    expect(content).toMatch(`props: /*#__PURE__*/mergeModels({ foo: String }`)
    expect(content).toMatch(`"modelValue": { default: 0 }`)
    expect(content).toMatch(
      `const count = useModel<any>(__ins.props, "modelValue")`
    )
    expect(content).not.toMatch('defineModel')
    expect(bindings).toStrictEqual({
      count: BindingTypes.SETUP_REF,
      foo: BindingTypes.PROPS,
      modelValue: BindingTypes.PROPS,
    })
  })

  test('w/ array props', () => {
    const { content, bindings } = compile(
      `
      <script setup>
      defineProps(['foo', 'bar'])
      const count = defineModel('count')
      </script>
    `,
      { defineModel: true }
    )
    assertCode(content)
    expect(content).toMatch(`props: /*#__PURE__*/mergeModels(['foo', 'bar'], {
    "count": {},
  })`)
    expect(content).toMatch(`const count = useModel<any>(__ins.props, "count")`)
    expect(content).not.toMatch('defineModel')
    expect(bindings).toStrictEqual({
      foo: BindingTypes.PROPS,
      bar: BindingTypes.PROPS,
      count: BindingTypes.SETUP_REF,
    })
  })

  test('w/ local flag', () => {
    const { content } = compile(
      `<script setup>
      const foo = defineModel({ local: true, default: 1 })
      const bar = defineModel('bar', { [key]: true })
      // const baz = defineModel('baz', { ...x })
      // const qux = defineModel('qux', x)

      // const foo2 = defineModel('foo2', { local: true, ...x })

      const local = true
      const hoist = defineModel('hoist', { local })
      </script>`,
      { defineModel: true }
    )
    assertCode(content)
    expect(content).toMatch(
      `useModel<any>(__ins.props, "modelValue", { local: true })`
    )
    expect(content).toMatch(
      `useModel<any>(__ins.props, "bar", { [key]: true })`
    )
    // expect(content).toMatch(`useModel(__ins.props, "baz", { ...x })`)
    // expect(content).toMatch(`useModel(__ins.props, "qux", x)`)
    // expect(content).toMatch(`useModel(__ins.props, "foo2", { local: true })`)
    expect(content).toMatch(`useModel<any>(__ins.props, "hoist", { local })`)
  })

  test('w/ types, basic usage', () => {
    const { content, bindings } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel<boolean | string>()
      const count = defineModel<number>('count')
      const disabled = defineModel<number>('disabled', { required: false })
      const any = defineModel<any | boolean>('any')
      const arr = defineModel<string[]>('arr')
      const arr1 = defineModel('arr1', { type: Array as PropType<string[]> })
      </script>
      `,
      { defineModel: true }
    )
    assertCode(content)
    expect(content).toMatch('"modelValue": { type: [Boolean, String] }')
    expect(content).toMatch('"count": { type: Number }')
    expect(content).toMatch(
      '"disabled": { type: Number, ...{ required: false } }'
    )
    expect(content).toMatch('"any": { type: [Object, Boolean] }')
    expect(content).toMatch('"arr": { type: Array as PropType<string[]> }')
    expect(content).toMatch('"arr1": { type: Array as PropType<string[]> }')
    expect(content).toMatch(
      'emits: ["update:modelValue", "update:count", "update:disabled", "update:any", "update:arr", "update:arr1"]'
    )

    expect(content).toMatch(
      `const modelValue = useModel<any>(__ins.props, "modelValue")`
    )
    expect(content).toMatch(
      `const count = useModel<number>(__ins.props, "count")`
    )
    expect(content).toMatch(
      `const disabled = useModel<number>(__ins.props, "disabled")`
    )
    expect(content).toMatch(`const any = useModel<any>(__ins.props, "any")`)
    expect(content).toMatch(
      `const arr = useModel<string[]>(__ins.props, "arr")`
    )
    expect(content).toMatch(
      `const arr1 = useModel<string[]>(__ins.props, "arr1")`
    )

    expect(bindings).toStrictEqual({
      modelValue: BindingTypes.SETUP_REF,
      count: BindingTypes.SETUP_REF,
      disabled: BindingTypes.SETUP_REF,
      any: BindingTypes.SETUP_REF,
      arr: BindingTypes.SETUP_REF,
      arr1: BindingTypes.SETUP_REF,
    })
  })

  test('w/ types, production mode', () => {
    const { content, bindings } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel<boolean>()
      const fn = defineModel<() => void>('fn')
      const fnWithDefault = defineModel<() => void>('fnWithDefault', { default: () => null })
      const str = defineModel<string>('str')
      const optional = defineModel<string>('optional', { required: false })
      </script>
      `,
      { defineModel: true, isProd: true }
    )
    assertCode(content)
    expect(content).toMatch('"modelValue": { type: Boolean }')
    expect(content).toMatch('"fn": { type: Function as PropType<() => void> }')
    expect(content).toMatch(
      '"fnWithDefault": { type: Function as PropType<() => void>, ...{ default: () => null } },'
    )
    expect(content).toMatch('"str": { type: String }')
    expect(content).toMatch(
      '"optional": { type: String, ...{ required: false } }'
    )
    expect(content).toMatch(
      'emits: ["update:modelValue", "update:fn", "update:fnWithDefault", "update:str", "update:optional"]'
    )
    expect(content).toMatch(
      `const modelValue = useModel<boolean>(__ins.props, "modelValue")`
    )
    expect(content).toMatch(
      `const fn = useModel<() => void>(__ins.props, "fn")`
    )
    expect(content).toMatch(`const str = useModel<string>(__ins.props, "str")`)
    expect(bindings).toStrictEqual({
      modelValue: BindingTypes.SETUP_REF,
      fn: BindingTypes.SETUP_REF,
      fnWithDefault: BindingTypes.SETUP_REF,
      str: BindingTypes.SETUP_REF,
      optional: BindingTypes.SETUP_REF,
    })
  })
})
