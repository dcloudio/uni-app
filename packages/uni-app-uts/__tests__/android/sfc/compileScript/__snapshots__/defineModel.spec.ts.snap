// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineModel() basic usage 1`] = `
"export default defineComponent({
  props: {
    "modelValue": { required: true },
    "count": {},
    "toString": { type: Function },
  },
  emits: ["update:modelValue", "update:count", "update:toString"],
  setup(__props): any | null {
const __ins = getCurrentInstance()!;
const _ctx = __ins.proxy;
const _cache = __ins.renderCache;

      const modelValue = useModel<any>(__ins.props, "modelValue")
      const c = useModel<any>(__ins.props, "count")
      const toString = useModel<Function>(__ins.props, "toString")
      
return (): any | null => { return null }
}

})"
`;

exports[`defineModel() w/ array props 1`] = `
"export default defineComponent({
  props: /*#__PURE__*/mergeModels(['foo', 'bar'], {
    "count": {},
  }),
  emits: ["update:count"],
  setup(__props): any | null {
const __ins = getCurrentInstance()!;
const _ctx = __ins.proxy;
const _cache = __ins.renderCache;

      
      const count = useModel<any>(__ins.props, "count")
      
return (): any | null => { return null }
}

})"
`;

exports[`defineModel() w/ defineProps and defineEmits 1`] = `
"export default defineComponent({
  props: /*#__PURE__*/mergeModels({ foo: String }, {
    "modelValue": { default: 0 },
  }),
  emits: /*#__PURE__*/mergeModels(['change'], ["update:modelValue"]),
  setup(__props): any | null {
const __ins = getCurrentInstance()!;
const _ctx = __ins.proxy;
const _cache = __ins.renderCache;

      
      
      const count = useModel<any>(__ins.props, "modelValue")
      
return (): any | null => { return null }
}

})"
`;

exports[`defineModel() w/ local flag 1`] = `
"const local = true
      
export default defineComponent({
  props: {
    "modelValue": { local: true, default: 1 },
    "bar": { [key]: true },
    "hoist": { local },
  },
  emits: ["update:modelValue", "update:bar", "update:hoist"],
  setup(__props): any | null {
const __ins = getCurrentInstance()!;
const _ctx = __ins.proxy;
const _cache = __ins.renderCache;

      const foo = useModel<any>(__ins.props, "modelValue", { local: true })
      const bar = useModel<any>(__ins.props, "bar", { [key]: true })
      // const baz = defineModel('baz', { ...x })
      // const qux = defineModel('qux', x)

      // const foo2 = defineModel('foo2', { local: true, ...x })

      const hoist = useModel<any>(__ins.props, "hoist", { local })
      
return (): any | null => { return null }
}

})"
`;

exports[`defineModel() w/ types, basic usage 1`] = `
"export default defineComponent({
  props: {
    "modelValue": { type: [Boolean, String] },
    "count": { type: Number },
    "disabled": { type: Number, ...{ required: false } },
    "any": { type: [Object, Boolean] },
    "arr": { type: Array as PropType<string[]> },
    "arr1": { type: Array as PropType<string[]> },
  },
  emits: ["update:modelValue", "update:count", "update:disabled", "update:any", "update:arr", "update:arr1"],
  setup(__props): any | null {
const __ins = getCurrentInstance()!;
const _ctx = __ins.proxy;
const _cache = __ins.renderCache;

      const modelValue = useModel<any>(__ins.props, "modelValue")
      const count = useModel<number>(__ins.props, "count")
      const disabled = useModel<number>(__ins.props, "disabled")
      const any = useModel<any>(__ins.props, "any")
      const arr = useModel<string[]>(__ins.props, "arr")
      const arr1 = useModel<string[]>(__ins.props, "arr1")
      
return (): any | null => { return null }
}

})"
`;

exports[`defineModel() w/ types, production mode 1`] = `
"export default defineComponent({
  props: {
    "modelValue": { type: Boolean },
    "fn": { type: Function as PropType<() => void> },
    "fnWithDefault": { type: Function as PropType<() => void>, ...{ default: () => null } },
    "str": { type: String },
    "optional": { type: String, ...{ required: false } },
  },
  emits: ["update:modelValue", "update:fn", "update:fnWithDefault", "update:str", "update:optional"],
  setup(__props): any | null {
const __ins = getCurrentInstance()!;
const _ctx = __ins.proxy;
const _cache = __ins.renderCache;

      const modelValue = useModel<boolean>(__ins.props, "modelValue")
      const fn = useModel<() => void>(__ins.props, "fn")
      const fnWithDefault = useModel<() => void>(__ins.props, "fnWithDefault")
      const str = useModel<string>(__ins.props, "str")
      const optional = useModel<string>(__ins.props, "optional")
      
return (): any | null => { return null }
}

})"
`;
